Sources/TensorFlow/Layers/Recurrent.swift

```diff
@@ -200,6 +200,80 @@ public struct LSTMCell<Scalar: TensorFlowFloatingPoint>: RNNCell {
     }
 }
 
+/// An GRU cell.
+public struct GRUCell<Scalar: TensorFlowFloatingPoint>: RNNCell {
+    public var updateWeight1, updateWeight2: Tensor<Scalar>
+    public var resetWeight1, resetWeight2: Tensor<Scalar>
+    public var outputWeight1, outputWeight2: Tensor<Scalar>
+    public var updateBias, outputBias, resetBias: Tensor<Scalar>
+
+    @noDerivative public var stateShape: TensorShape {
+        [1, updateWeight.shape[0]]
+    }
+
+    public var zeroState: State {
+        State(hidden: Tensor(zeros: stateShape))
+    }
+
+    public typealias TimeStepInput = Tensor<Scalar>
+    public typealias TimeStepOutput = State
+    public typealias Input = RNNCellInput<TimeStepInput, State>
+    public typealias Output = RNNCellOutput<TimeStepOutput, State>
+
+    /// Creates a `GRUCell` with the specified input size and hidden state size.
+    ///
+    /// - Parameters:
+    ///   - inputSize: The number of features in 2-D input tensors.
+    ///   - hiddenSize: The number of features in 2-D hidden states.
+    public init(
+        inputSize: Int,
+        hiddenSize: Int,
+        seed: TensorFlowSeed = Context.local.randomSeed
+    ) {
+        let gateWeightShape = TensorShape([inputSize, 1])
+        let gateBiasShape = TensorShape([hiddenSize])
+        self.updateWeight1 = Tensor(glorotUniform: gateWeightShape, seed: seed)
+        self.updateWeight2 = Tensor(glorotUniform: gateWeightShape, seed: seed)
+        self.updateBias = Tensor(zeros: gateBiasShape)
+        self.resetWeight1 = Tensor(glorotUniform: gateWeightShape, seed: seed)
+        self.resetWeight2 = Tensor(glorotUniform: gateWeightShape, seed: seed)
+        self.resetBias = Tensor(zeros: gateBiasShape)
+        self.outputWeight1 = Tensor(glorotUniform: gateWeightShape, seed: seed)
+        self.outputWeight2 = Tensor(glorotUniform: gateWeightShape, seed: seed)
+        self.outputBias = Tensor(zeros: gateBiasShape)
+    }
+
+    public struct State: Differentiable {
+        public var hidden: Tensor<Scalar>
```

Comment:
Given that `State` is just a single `Tensor<Scalar>` why don't you use that directly and define a new struct?

Reply:
Sorry, I need more explanation. 
> why don't you use that directly
So make `hidden` is a property of `GRUCell` instead of `State`?
> and define a new struct
No idea what you mean. If hidden is used directly, then State is empty now. Is that what you are talking about?

Reply:
You can make `State` be simply a typealias for `Tensor<Scalar>`. In this case you would not need this struct at all and you would also have to change `public var zeroState: State` to `public var zeroState: Tensor<Scalar>`.

Reply:
Did you give this a shot is there a limitation that forces you to create a wrapping struct?

Reply:
@eaplatanios Sorry for the delay, I have been very busy lately. I will take a look this week. Unfortunately, I am very new to this repo, so this might take some time.

Reply:
@eaplatanios made an update [here](https://github.com/tensorflow/swift-apis/pull/417/commits/97dd831a353e3324e4787a27178ae293dd98e7ed). Please let me know if this is what you had in mind.

Everything looks correct to me. This builds and passes the tests, but at the end of the `swift build` or `swift test` command, there is an assertion failure.

<details>
  <summary>Error Message</summary>
Assertion failed: (rootWTable && "root witness table not bound in local context!"), function emitArchetypeWitnessTableRef, file /Users/swiftninjas/s4tf/swift/lib/IRGen/GenArchetype.cpp, line 245.
Stack dump:
0.	Program arguments: /Library/Developer/Toolchains/swift-tensorflow-RELEASE-0.5.xctoolchain/usr/bin/swift -frontend -c /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/BackwardsCompatibility.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Bindings/EagerExecution.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Bindings/RawOpsGenerated.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Bindings/TFTensorOperation.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Context.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/ArrayOps.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/DataTypes.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/DifferentialOperators.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/Execution.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/LazyTensorContext.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/LazyTensorOperation.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/LazyTensorShapeInference.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/LazyTensorTFFunctionBuilder.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/LazyTensorTrace.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/LazyTensorTraceCache.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/PythonConversion.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/Runtime.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/Serialization.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/ShapedArray.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/StringTensor.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/Tensor.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/TensorGroup.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/TensorHandle.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/TensorShape.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/Threading.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Core/Utilities.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Exports.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Initializers.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Layer.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Layers/Convolutional.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Layers/Core.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Layers/Dense.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Layers/Dropout.swift -primary-file /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Layers/Embedding.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Layers/Initialization.swift -primary-file /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Layers/Normalization.swift -primary-file /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Layers/Pooling.swift -primary-file /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Layers/Recurrent.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Layers/Sequential.swift -primary-file /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Layers/Upsampling.swift -primary-file /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Loss.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Operators/Basic.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Operators/Comparison.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Operators/Dataset.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Operators/Math.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Operators/NN.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Optimizers/MomentumBased.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Optimizers/Optimizer.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Optimizers/SGD.swift /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/StdlibExtensions.swift -supplementary-output-file-map /var/folders/9g/3wb_r1j52jj8bp_ld6sppb300000gn/T/supplementaryOutputs-f58827 -target x86_64-apple-macosx10.10 -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -I /Users/devinhaslam/Documents/tmp/swift-apis/.build/x86_64-apple-macosx/debug -F /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -enable-testing -g -module-cache-path /Users/devinhaslam/Documents/tmp/swift-apis/.build/x86_64-apple-macosx/debug/ModuleCache -swift-version 4.2 -Onone -D SWIFT_PACKAGE -D DEBUG -color-diagnostics -enable-anonymous-context-mangled-names -parse-as-library -module-name TensorFlow -o /Users/devinhaslam/Documents/tmp/swift-apis/.build/x86_64-apple-macosx/debug/TensorFlow.build/Layers/Embedding.swift.o -o /Users/devinhaslam/Documents/tmp/swift-apis/.build/x86_64-apple-macosx/debug/TensorFlow.build/Layers/Normalization.swift.o -o /Users/devinhaslam/Documents/tmp/swift-apis/.build/x86_64-apple-macosx/debug/TensorFlow.build/Layers/Pooling.swift.o -o /Users/devinhaslam/Documents/tmp/swift-apis/.build/x86_64-apple-macosx/debug/TensorFlow.build/Layers/Recurrent.swift.o -o /Users/devinhaslam/Documents/tmp/swift-apis/.build/x86_64-apple-macosx/debug/TensorFlow.build/Layers/Upsampling.swift.o -o /Users/devinhaslam/Documents/tmp/swift-apis/.build/x86_64-apple-macosx/debug/TensorFlow.build/Loss.swift.o -index-store-path /Users/devinhaslam/Documents/tmp/swift-apis/.build/x86_64-apple-macosx/debug/index/store -index-system-modules 
1.	Swift version 5.1-dev (LLVM 200186e28b, Swift 1238976565)
2.	While emitting IR SIL function "@AD__$s10TensorFlow7GRUCellV14callAsFunctionyAA13RNNCellOutputVyAA0A0VyxGAIGAA0G5InputVyA2IGF__vjp_src_0_wrt_1".
 for 'callAsFunction(_:)' (at /Users/devinhaslam/Documents/tmp/swift-apis/Sources/TensorFlow/Layers/Recurrent.swift:313:12)
0  swift                    0x000000010b990715 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  swift                    0x000000010b98fa05 llvm::sys::RunSignalHandlers() + 85
2  swift                    0x000000010b990cf8 SignalHandler(int) + 264
3  libsystem_platform.dylib 0x00007fff72844b1d _sigtramp + 29
4  libsystem_platform.dylib 0x00000000000045c8 _sigtramp + 2373712584
5  libsystem_c.dylib        0x00007fff7271aa1c abort + 120
6  libsystem_c.dylib        0x00007fff72719cd6 err + 0
7  swift                    0x000000010ba36fe3 swift::irgen::emitArchetypeWitnessTableRef(swift::irgen::IRGenFunction&, swift::CanTypeWrapper<swift::ArchetypeType>, swift::ProtocolDecl*) (.cold.6) + 35
8  swift                    0x00000001080670fd swift::irgen::emitArchetypeWitnessTableRef(swift::irgen::IRGenFunction&, swift::CanTypeWrapper<swift::ArchetypeType>, swift::ProtocolDecl*) + 1165
9  swift                    0x000000010813072d swift::irgen::emitWitnessTableRef(swift::irgen::IRGenFunction&, swift::CanType, llvm::Value**, swift::ProtocolConformanceRef) + 189
10 swift                    0x00000001081338ce bool llvm::function_ref<bool (unsigned int, swift::CanType, swift::ProtocolDecl*)>::callback_fn<emitConditionalConformancesBuffer(swift::irgen::IRGenFunction&, swift::ProtocolConformance const*)::$_15>(long, unsigned int, swift::CanType, swift::ProtocolDecl*) + 174
11 swift                    0x000000010891026d swift::SILWitnessTable::enumerateWitnessTableConditionalConformances(swift::ProtocolConformance const*, llvm::function_ref<bool (unsigned int, swift::CanType, swift::ProtocolDecl*)>) + 221
12 swift                    0x00000001081335b4 emitWitnessTableAccessorCall(swift::irgen::IRGenFunction&, swift::ProtocolConformance const*, llvm::Value**) + 260
13 swift                    0x00000001081307a4 swift::irgen::emitWitnessTableRef(swift::irgen::IRGenFunction&, swift::CanType, llvm::Value**, swift::ProtocolConformanceRef) + 308
14 swift                    0x0000000108131262 swift::irgen::emitGenericRequirementFromSubstitutions(swift::irgen::IRGenFunction&, swift::CanGenericSignature, swift::ModuleDecl&, swift::irgen::GenericRequirement, swift::SubstitutionMap) + 178
15 swift                    0x0000000108135a32 void llvm::function_ref<void (swift::irgen::GenericRequirement)>::callback_fn<(anonymous namespace)::EmitPolymorphicArguments::emit(swift::SubstitutionMap, swift::irgen::WitnessMetadata*, swift::irgen::Explosion&)::$_21>(long, swift::irgen::GenericRequirement) + 34
16 swift                    0x0000000108135b23 void llvm::function_ref<void (swift::irgen::GenericRequirement)>::callback_fn<(anonymous namespace)::PolymorphicConvention::enumerateUnfulfilledRequirements(llvm::function_ref<void (swift::irgen::GenericRequirement)> const&)::$_11>(long, swift::irgen::GenericRequirement) + 211
17 swift                    0x000000010812a0d5 swift::irgen::enumerateGenericSignatureRequirements(swift::CanGenericSignature, llvm::function_ref<void (swift::irgen::GenericRequirement)> const&) + 277
18 swift                    0x0000000108130932 swift::irgen::emitPolymorphicArguments(swift::irgen::IRGenFunction&, swift::CanTypeWrapper<swift::SILFunctionType>, swift::SubstitutionMap, swift::irgen::WitnessMetadata*, swift::irgen::Explosion&) + 306
19 swift                    0x00000001080f117d emitPartialApplicationForwarder(swift::irgen::IRGenModule&, llvm::Optional<swift::irgen::FunctionPointer> const&, bool, swift::irgen::Signature const&, swift::CanTypeWrapper<swift::SILFunctionType>, swift::CanTypeWrapper<swift::SILFunctionType>, swift::CanTypeWrapper<swift::SILFunctionType>, swift::SubstitutionMap, swift::irgen::HeapLayout const*, llvm::ArrayRef<swift::ParameterConvention>) + 6557
20 swift                    0x00000001080ef22a swift::irgen::emitFunctionPartialApplication(swift::irgen::IRGenFunction&, swift::SILFunction&, swift::irgen::FunctionPointer const&, llvm::Value*, swift::irgen::Explosion&, llvm::ArrayRef<swift::SILParameterInfo>, swift::SubstitutionMap, swift::CanTypeWrapper<swift::SILFunctionType>, swift::CanTypeWrapper<swift::SILFunctionType>, swift::CanTypeWrapper<swift::SILFunctionType>, swift::irgen::Explosion&, bool) + 4250
21 swift                    0x000000010819c40a swift::SILInstructionVisitor<(anonymous namespace)::IRGenSILFunction, void>::visit(swift::SILInstruction*) + 53226
22 swift                    0x000000010818c223 swift::irgen::IRGenModule::emitSILFunction(swift::SILFunction*) + 8403
23 swift                    0x00000001080a8a6e swift::irgen::IRGenerator::emitGlobalTopLevel() + 734
24 swift                    0x0000000108166b92 performIRGeneration(swift::IRGenOptions&, swift::ModuleDecl*, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::StringRef, swift::PrimarySpecificPaths const&, llvm::LLVMContext&, swift::SourceFile*, llvm::GlobalVariable**) + 1282
25 swift                    0x0000000108167082 swift::performIRGeneration(swift::IRGenOptions&, swift::SourceFile&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::StringRef, swift::PrimarySpecificPaths const&, llvm::LLVMContext&, llvm::GlobalVariable**) + 82
26 swift                    0x000000010801385e performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 13294
27 swift                    0x000000010800f540 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3024
28 swift                    0x0000000107fb6c59 main + 729
29 libdyld.dylib            0x00007fff726432e5 start + 1

</details>

Thank you!

Reply:
https://bugs.swift.org/browse/SR-10697 might be the issue. Applying a workaround similar to https://github.com/tensorflow/swift-apis/pull/123 may fix it.

Reply:
@marcrasi That seems to be the problem exactly. Do you think that I should add a TODO or just wait until the fix is implemented?

Reply:
I suggest changing it back to something that works and add a TODO. That bug has been inactive for a while, so I don't expect that it'll be fixed any time soon.

Reply:
@marcrasi All done 😄